File: ios/MunderMobile/Sources/RemoteCrypto.swift
import Foundation
import CryptoKit
import Security

/// munder-remote@1, the same wire the office speaks (tools/munder/lib-remote.cjs)
/// and the PWA speaks (tools/munder/remote-app/remote-crypto.js). Everything here
/// is CryptoKit: X25519, HKDF-SHA256, ChaCha20-Poly1305, SHA-256. Tests/vectors.json
/// is computed by the office's own Node code and RemoteCryptoTests must match it
/// byte for byte.
public enum RemoteCrypto {
    public static let protocolName = "munder-remote@1"

    /// First 8 bytes of SHA-256(raw X25519 public key), as 16 hex chars.
    public static func deviceId(publicKey: Data) -> String {
        SHA256.hash(data: publicKey).prefix(8).map { String(format: "%02x", $0) }.joined()
    }

    /// The 6 digits both screens show. The phone's nonce was committed before the office answered.
    public static func sas(officeBoxPub: String, devicePub: String, officeNonce: String, deviceNonce: String) -> String {
        let text = "\(protocolName)|sas|\(officeBoxPub)|\(devicePub)|\(officeNonce)|\(deviceNonce)"
        let b = Array(SHA256.hash(data: Data(text.utf8)))
        let n = (UInt32(b[0]) << 24) | (UInt32(b[1]) << 16) | (UInt32(b[2]) << 8) | UInt32(b[3])
        return String(format: "%06u", n % 1_000_000)
    }

    /// What the phone sends first: the hash of its nonce, never the nonce.
    public static func commit(nonce: Data) -> String {
        Data(SHA256.hash(data: nonce)).base64URL
    }

    public static func sessionKey(devicePrivate: Curve25519.KeyAgreement.PrivateKey, officeBoxPub: String, officeId: String, deviceId: String) throws -> SymmetricKey {
        guard let raw = Data(base64URL: officeBoxPub), raw.count == 32 else { throw RemoteError.badOffice }
        let pub = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: raw)
        let secret = try devicePrivate.sharedSecretFromKeyAgreement(with: pub)
        return secret.hkdfDerivedSymmetricKey(
            using: SHA256.self,
            salt: Data("\(officeId)|\(deviceId)".utf8),
            sharedInfo: Data("\(protocolName) key".utf8),
            outputByteCount: 32
        )
    }

    static func aad(_ direction: String, deviceId: String, officeId: String) -> Data {
        Data("\(protocolName)|\(direction)|\(deviceId)|\(officeId)".utf8)
    }

    /// ciphertext ‖ 16-byte tag, exactly what Node's chacha20-poly1305 produces.
    public static func seal(_ plaintext: Data, key: SymmetricKey, iv: Data, direction: String, deviceId: String, officeId: String) throws -> Data {
        let box = try ChaChaPoly.seal(
            plaintext,
            using: key,
            nonce: ChaChaPoly.Nonce(data: iv),
            authenticating: aad(direction, deviceId: deviceId, officeId: officeId)
        )
        return Data(box.ciphertext) + Data(box.tag)
    }

    public static func open(_ sealed: Data, key: SymmetricKey, iv: Data, direction: String, deviceId: String, officeId: String) throws -> Data {
        guard sealed.count >= 16 else { throw RemoteError.badReply }
        let bytes = Data(sealed)
        let box = try ChaChaPoly.SealedBox(
            nonce: ChaChaPoly.Nonce(data: iv),
            ciphertext: Data(bytes.prefix(bytes.count - 16)),
            tag: Data(bytes.suffix(16))
        )
        return try ChaChaPoly.open(box, using: key, authenticating: aad(direction, deviceId: deviceId, officeId: officeId))
    }

    public static func randomBytes(_ count: Int) -> Data {
        var data = Data(count: count)
        let status = data.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, count, $0.baseAddress!) }
        precondition(status == errSecSuccess, "no randomness")
        return data
    }
}

public extension Data {
    init?(base64URL: String) {
        var s = base64URL.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        while s.count % 4 != 0 { s += "=" }
        self.init(base64Encoded: s)
    }

    var base64URL: String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

public enum RemoteError: Error, LocalizedError, Equatable {
    /// The office answered with an error before sealing (not paired, waiting, rate limited…).
    case http(status: Int, code: String, message: String)
    /// The sealed answer said ok:false.
    case remote(code: String, message: String)
    case badReply
    case badOffice
    case unreachable(String)
    case notPaired

    public var code: String {
        switch self {
        case .http(_, let code, _), .remote(let code, _): return code
        case .badReply: return "bad_reply"
        case .badOffice: return "bad_office"
        case .unreachable: return "unreachable"
        case .notPaired: return "not_paired"
        }
    }

    public var errorDescription: String? {
        switch self {
        case .http(_, _, let message), .remote(_, let message): return message
        case .badReply: return "La respuesta de la oficina no corresponde a esta llamada."
        case .badOffice: return "Esa no parece una oficina Munder."
        case .unreachable(let why): return "No alcanzo la oficina (\(why)). ¿Está encendido el enlace y estás en tu red o en Tailscale?"
        case .notPaired: return "Este celular no está emparejado."
        }
    }
}

File: android/MunderMobile/app/src/main/java/mx/isyco/munder/mobile/RemoteCrypto.kt
package mx.isyco.munder.mobile

import android.util.Base64
import java.nio.ByteBuffer
import java.security.MessageDigest
import java.security.SecureRandom
import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.generators.HKDFBytesGenerator
import org.bouncycastle.crypto.modes.ChaCha20Poly1305
import org.bouncycastle.crypto.params.AEADParameters
import org.bouncycastle.crypto.params.HKDFParameters
import org.bouncycastle.crypto.params.KeyParameter
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import org.bouncycastle.crypto.params.X25519PublicKeyParameters

/**
 * munder-remote@1, el mismo cable que habla la oficina (tools/munder/lib-remote.cjs)
 * y que habla iOS (Sources/RemoteCrypto.swift).
 *
 * Puerto línea por línea del Swift con CryptoKit a BouncyCastle puro-Java:
 * X25519, HKDF-SHA256, ChaCha20-Poly1305, SHA-256.
 * Tests/vectors.json lo calcula el Node de la oficina; el test Android debe
 * dar los mismos bytes que el test Swift.
 */
object RemoteCrypto {
    const val PROTOCOL = "munder-remote@1"

    private val rng = SecureRandom()

    fun randomBytes(count: Int): ByteArray = ByteArray(count).also { rng.nextBytes(it) }

    /** Primeros 8 bytes de SHA-256(llave pública X25519 cruda), 16 hex. */
    fun deviceId(publicKeyRaw: ByteArray): String {
        val h = MessageDigest.getInstance("SHA-256").digest(publicKeyRaw)
        return h.take(8).joinToString("") { "%02x".format(it) }
    }

    /** Los 6 dígitos que muestran las dos pantallas. El nonce del móvil se comprometió antes. */
    fun sas(officeBoxPub: String, devicePub: String, officeNonce: String, deviceNonce: String): String {
        val text = "$PROTOCOL|sas|$officeBoxPub|$devicePub|$officeNonce|$deviceNonce"
        val h = MessageDigest.getInstance("SHA-256").digest(text.toByteArray(Charsets.UTF_8))
        val n = ByteBuffer.wrap(h, 0, 4).int.toLong() and 0xFFFFFFFFL
        return "%06d".format((n % 1_000_000L).toInt())
    }

    /** Lo primero que manda el móvil: el hash de su nonce, nunca el nonce. */
    fun commit(nonce: ByteArray): String {
        val h = MessageDigest.getInstance("SHA-256").digest(nonce)
        return h.toB64u()
    }

    data class KeyPair(val privateRaw: ByteArray, val publicRaw: ByteArray)

    fun generateKeyPair(): KeyPair {
        val priv = X25519PrivateKeyParameters(rng)
        val pub = X25519PublicKeyParameters(priv.encoded, 0)
        return KeyPair(priv.encoded, pub.encoded)
    }

    fun publicOf(privateRaw: ByteArray): ByteArray {
        val priv = X25519PrivateKeyParameters(privateRaw, 0)
        return X25519PublicKeyParameters(priv.encoded, 0).encoded
    }

    /** Llave de sesión: X25519 + HKDF-SHA256(salt="office|device", info="munder-remote@1 key"). */
    fun sessionKey(devicePrivateRaw: ByteArray, officeBoxPubB64u: String, officeId: String, deviceId: String): ByteArray {
        val officeRaw = officeBoxPubB64u.fromB64u()
            ?: throw RemoteError.BadOffice()
        require(officeRaw.size == 32) { "llave de oficina inválida" }
        val priv = X25519PrivateKeyParameters(devicePrivateRaw, 0)
        val pub = X25519PublicKeyParameters(officeRaw, 0)
        val secret = ByteArray(32)
        // BouncyCastle ≥1.72: el acuerdo se calcula con agreement.calculateAgreement.
        val agreement = org.bouncycastle.crypto.agreement.X25519Agreement()
        agreement.init(priv)
        agreement.calculateAgreement(pub, secret, 0)
        val hkdf = HKDFBytesGenerator(SHA256Digest())
        hkdf.init(
            HKDFParameters(
                secret,
                "$officeId|$deviceId".toByteArray(Charsets.UTF_8),
                "$PROTOCOL key".toByteArray(Charsets.UTF_8),
            )
        )
        val out = ByteArray(32)
        hkdf.generateBytes(out, 0, 32)
        return out
    }

    fun aad(direction: String, deviceId: String, officeId: String): ByteArray =
        "$PROTOCOL|$direction|$deviceId|$officeId".toByteArray(Charsets.UTF_8)

    /** ciphertext ‖ tag de 16, exactamente lo que produce Node chacha20-poly1305. */
    fun seal(plaintext: ByteArray, key: ByteArray, iv: ByteArray, direction: String, deviceId: String, officeId: String): ByteArray {
        val aead = ChaCha20Poly1305()
        aead.init(true, AEADParameters(KeyParameter(key), 128, iv, aad(direction, deviceId, officeId)))
        val out = ByteArray(aead.getOutputSize(plaintext.size))
        var off = aead.processBytes(plaintext, 0, plaintext.size, out, 0)
        off += aead.doFinal(out, off)
        return out.copyOf(off)
    }

    fun open(sealed: ByteArray, key: ByteArray, iv: ByteArray, direction: String, deviceId: String, officeId: String): ByteArray {
        if (sealed.size < 16) throw RemoteError.BadReply()
        val aead = ChaCha20Poly1305()
        aead.init(false, AEADParameters(KeyParameter(key), 128, iv, aad(direction, deviceId, officeId)))
        val out = ByteArray(aead.getOutputSize(sealed.size))
        return try {
            var off = aead.processBytes(sealed, 0, sealed.size, out, 0)
            off += aead.doFinal(out, off)
            out.copyOf(off)
        } catch (e: Exception) {
            throw RemoteError.BadReply()
        }
    }

    // ── base64url sin relleno, igual que Node Buffer.toString('base64url') ──

    fun ByteArray.toB64u(): String =
        Base64.encodeToString(this, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)

    fun String.fromB64u(): ByteArray? = try {
        var s = this.replace('-', '+').replace('_', '/')
        while (s.length % 4 != 0) s += "="
        Base64.decode(s, Base64.DEFAULT)
    } catch (e: Exception) {
        null
    }
}

/** Mismos códigos que RemoteError en Swift: la UI los traduce igual. */
sealed class RemoteError(message: String) : Exception(message) {
    open val code: String = "error"
    class Http(val status: Int, override val code: String, msg: String) : RemoteError(msg)
    class Remote(override val code: String, msg: String) : RemoteError(msg)
    class BadReply : RemoteError("La respuesta de la oficina no corresponde a esta llamada.") {
        override val code = "bad_reply"
    }
    class BadOffice : RemoteError("Esa no parece una oficina Munder.") {
        override val code = "bad_office"
    }
    class Unreachable(why: String) : RemoteError("No alcanzo la oficina ($why). ¿Está encendido el enlace y estás en tu red o en Tailscale?") {
        override val code = "unreachable"
    }
    class NotPaired : RemoteError("Este celular no está emparejado.") {
        override val code = "not_paired"
    }
}
